Add a plugin tutorial (#2264)
jupyter-book/jupyter-book: docs/tutorial/plugins.md

---
title: Create Plugins in JavaScript
description: Plugins implemented in JavaScript are easily used across different projects, as they do not require any additional programs to be installed.
---

Jupyter Book uses the [MyST plugins infrastructure](xref:guide/plugins) to extend Jupyter Book's base functionality.
Here are some things you can do with plugins:

- Define a new directive or role to generate block-level or inline content.
- Define a transform that modifies a page's content at build time.

Plugins are natively written in JavaScript, or by writing an executable script in another language that follows [the MyST plugin protocol](xref:guide/executable-plugins).
This tutorial shows you how to create a JavaScript plugin from scratch.

## Create a basic Jupyter Book template

We'll start by creating a basic template book that we can add to.

🛠 First, create a new folder for a demo Jupyter Book (or use a pre-existing book).

```shell
$ mkdir plugin-demo
$ cd plugin-demo
```

🛠 Next initialize a Jupyter Book in the folder to add a configuration file:

```shell

> $ jupyter-book init

Reviewer: In JB, we are currently leaning in to the `jupyter book` dispatch mechanism, so we should change this.

```

🛠 Finally, create an empty file that we'll use to play around with our plugin:

> ```shell
> $ echo '# My demo page' > myplugindemo.md
> ```

Reviewer: ````suggestion
```{code} markdown
:filename: myplugindemo.md

# My demo page
```
````

I think prefer platform agnostic instructions?

🛠 And add it to your Jupyter Book's configuration file:

```yaml
project:
  toc:
    - file: myplugindemo.md
```

> You should now be able to run `myst start` and see a page with just your title rendered:

Reviewer: ```suggestion
You should now be able to run `jupyter book start` and see a page with just your title rendered:
```

> 🛠️ Run `myst start` to confirm that your demo book is set up properly.

Reviewer: ```suggestion
🛠️ Run `jupyter book start` to confirm that your demo book is set up properly.
```

```shell

> $ myst start

Reviewer: ```suggestion
$ jupyter book start
```

```

## Create an empty plugin script

Next we'll create an empty JavaScript file[^esm] that we'll use to add new plugin logic.

> [^esm]: The format of the Javascript should be an ECMAScript modules, not CommonJS. This means it uses `import` statements rather than `require()` and is the most modern style of Javascript. See [the MDN modules documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) for background on modules.

Reviewer: ```suggestion
[^esm]: The format of the JavaScript should be an ECMAScript modules, not CommonJS. This means it uses `import` statements rather than `require()` and is the most modern style of JavaScript. See [the MDN modules documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) for background on modules.
```

🛠 Create a `src/` folder and add an empty JavaScript file to it

> ```shell
> $ mkdir src
> $ touch src/myplugin.mjs
> ```

Reviewer: ```suggestion
```{code}
:filename: src/myplugin.mjs

```
````

In this tutorial we will define a **directive** and play around with ways to control it.
Other plugins you could define are **roles** and **transforms**.

### Register the plugin

You register plugins by defining JavaScript objects that follow the **plugin specification**[^esm] - this is a way of structuring plugin configuration so that Jupyter Book and MyST know how to read them.

Below we'll create the skeleton of a MyST plugin - the following defines an empty plugin structure and "exports" the plugin so that MyST can use it.
We'll add the specific code later.

🛠 Copy and paste the JavaScript below into the `myplugin.mjs` file that you created.

```{code} javascript
:filename: src/myplugin.mjs
const plugin = {
  name: "My cool plugin",
  directives: [],
};

export default plugin;
```

We'll register the plugin so that MyST can discover and use it.

🛠 Add the following to your `myst.yml` file:

```yaml
project:
  plugins:
    - src/myplugin.mjs
```

🛠 Start your MyST site.

```shell

> $ myst start

Reviewer: ```suggestion
$ jupyter book start
```

  },
};
```

In this directive, we are **constructing MyST {term}`AST`** using the directive's argument.
MyST AST is the structure built when you "build a MyST site".
Above, we're defining two MyST nodes: a `paragraph` node (`type: "paragraph"`) and a `text` node (`type: "text"`). We're then nesting the `text` node underneath the `paragraph` node (`children: [sentence]`).

Finally, we **return an array of MyST nodes**.
This array is what will be inserted into the document when the directive is run.
You must always return an array (list) of nodes in a directive or role.

Next we add this directive to the list of plugin exports for our plugin:

🛠️ Modify the plugin exports so that it looks like the following:

```{code} javascript
:filename: myplugindemo.md

const plugin = {
  name: "My cool plugin",
  directives: [myDirective],
};
```

Now our plugin should be ready to use!
We'll add it to our content to try it out.

🛠️ Add the directive to your MyST document.

````{code} markdown
:filename: myplugindemo.md
# My demo page

```{mydirective} myword
```
````

🛠️ **Re-build** MyST (you must re-build manually any time you change a plugin)

```shell

> $ myst start

Reviewer: ```suggestion
$ jupyter book start
```

Transforms can be used to modify the AST of a page, so let's demonstrate this by adding **a footer** to each card produced by our directives.
It will display the total number of times that directive has been used on a page.

🛠️ Create the skeleton of a Transform and register it in your plugin output like so:

```{code} javascript
:filename: src/myplugin.mjs

const myTransform = {
  name: "directive-counter",
  doc: "Count the number of word directives on a page.",
  stage: "document",
  plugin: (_, utils) => (node) => {
  },
};

const plugin = {
  name: "My cool plugin",
  directives: [myDirective],
  transforms: [myTransform],
};
```

The logic of a transform lives in the `plugin:` value.
It should be a function that operates on nodes (see the structure above).
Each node is a MyST AST for a page.

The `utils` parameter is a helper module to provide some functions for manipulating MyST AST.
We'll use it in our plugin to find all of our directive nodes.

First, let's confirm that this plugin is properly loaded.

🛠️ Re-start MyST to confirm that the plugin's transform is loaded.

```shell

> $ myst start

Reviewer: ```suggestion
$ jupyter book start
```

            }],
        });
    });
  }
};
```

Here's a short description of what it does:

- `plugin:` expects a _function_ that is called on the document tree (the AST of the page). That's what the first line defines.
- Use the `utils` module to find all of the nodes for a `card` on the page. `utils` has a few helper functions to make it easier to create plugins.
- Loop through each card node and add footer content to its AST.[^footer]

[^footer]: Not sure what the AST for footer looks like? Try [this MyST sandbox demo](https://mystmd.org/sandbox?tab=demo&myst=YABgAGAAewBjAGEAcgBkAH0AIABNAHkAIAB0AGkAdABsAGUACgA6AGgAZQBhAGQAZQByADoAIABNAHkAIABoAGUAYQBkAGUAcgAKADoAZgBvAG8AdABlAHIAOgAgAE0AeQAgAGYAbwBvAHQAZQByAAoATQB5ACAAYgBvAGQAeQAuAAoAYABgAGAA).

Let's add bodies to each of the cards on our markdown, and then re-build to see the result.

🛠️ Modify your markdown file to look like this.

````{code} markdown
:filename: myplugindemo.md
# My demo page

```{mydirective} Not bold
A body!
```

```{mydirective} Bold
:bold: true
Another body!
```

```{mydirective} A card
Here's a card body!
```
````

🛠️ Re-build your demo site.

```shell

> $ myst start

Reviewer: ```suggestion
$ jupyter book start
```

- **name**: The name of the directive in MyST (ie, how you will write the direcive)
- **doc**: A docstring to help others know what the directive does.
- **arg**: Tell MyST that this directive expects an argument of a "String" type.
- **`run(data, vfile, ctx)`**: A special function all directives must have. This is executed when the directive is run.
  - **`data`**: A data object provided to each directive when it is run. This contains information about the directive, and also other information about what's on the page.
  - **`vfile`**: A [VFile](https://github.com/vfile/vfile) object. This is a virtual file object that contains information about the page that the directive is being run on.

>   - **`ctx`**: A context object provided to each directive when it is run. This provides helper functions and information for working with MyST.

Reviewer: Tempted to use a deflist here.
